feature/client

elmigranto opened the pull request:
`Client` from #3.

Still left (feel free to add stuff):

  - [ ] pushing events
  - [x] supprot for `after` and `limit` (right now, paging is `clientID` based, so #7 is needed)
  - [ ] [FOR LATER] multiple requests are running when 1 could've been enough (listening on "master" like `service` and "sub" like `service:user-added` — 2 requests, but 1 could've been enough)

  - channel can not be named some things:
    ``` js
    // Some events are special, and have nothing to do with our channles,
    // we should not start HTTP request for them. should not monitor them:
    const ignoreChannels = [
      'newListener',    // EventEmitter's after listener added
      'removeListener', // EventEmitter's after listener removed
      'error',          // HTTP request errored
      'drain'           // All the requests are finished and there are no listeners left.
    ];
    ```
  - see #7 about ack

elmigranto committed:
Implement Client with ability to listen for events.

elmigranto committed:
Export appropriately for `require(‘ganomede-events’)`.

elmigranto committed:
Document `Client` API more in-depth.

elmigranto committed:
`setenv` does not export `API_SECRET` (but `npm test` does).

elmigranto commented:
About those double-events and "*useless channels* vs *useless messages*" — I guess double emit of same event is desired, since technically we push same message to multiple channels… 

The solution is probably not emit to "master channel" when subchannel receives a message in client. That leaves us with somewhat useless HTTP(S) request when both, master and sub, are listened to, but that's different problem.

elmigranto committed:
Do not categorize channels in “master” / “sub” — just `emit` to what was asked of us.

elmigranto commented:
All right, changed it to emit `eventName` of exactly channel that was asked to be listened on. Still have multiple requests going, though, when 1 could've been enough.

j3k0 commented:
If polling is ready, then you can merge and `npm publish` the initial version, because then the work on mailchimp worker can start. You can make another PR with the remaining requirements.

elmigranto commented:
Yeah, sure, let's do that.

> On 14 Feb 2017, at 21:44, Jean-Christophe Hoelt <notifications@github.com> wrote:
> 
> If polling is ready, then you can merge and npm publish the initial version, because then the work on mailchimp worker can start. You can make another PR with the remaining requirements.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

elmigranto commented:
Though we'll have to do clientId for server or after for client first. 

> On 14 Feb 2017, at 21:44, Jean-Christophe Hoelt <notifications@github.com> wrote:
> 
> If polling is ready, then you can merge and npm publish the initial version, because then the work on mailchimp worker can start. You can make another PR with the remaining requirements.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

elmigranto committed:
Client: support `after`.

elmigranto committed:
Separate HTTP parsing and event polling to its own functions.

elmigranto committed:
Drop useless tests and redo `EventsStore`:
  - no need to check redis implementation everywhere;
  - no need to validate inputs on every level (HTTP endpoint is enough).

elmigranto committed:
Added `parsePostParams` for validating input when adding event.

elmigranto committed:
Redo RedisStore a bit:
  - it now relies on `message.id` instead of filling it by itself;
  - drop `fakeredis` (does not implement commands correctly, use `redis` instead);
  - fast tests :)

elmigranto committed:
Added `redis` to CI.

elmigranto committed:
Close redis connection after tests are done.

elmigranto committed:
Make CI use Redis v3 instead of pre-installed older one.

elmigranto committed:
Changed signature of `EventsStore#loadEvents()` to `(channel, pagingParams, callback)`.

elmigranto committed:
Renamed `RedisStore#getIndex()` to `#nextIndex()`.

elmigranto committed:
Implemented retrieving `after` when it is not explicitly set based on `clientId`.

elmigranto committed:
Update `last-fetched` index of a client on GET requests with explicit `after` param.
